RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc

```diff
@@ -0,0 +1,126 @@
+#include "RecoTracker/FinalTrackSelectors/interface/TrackMVAClassifier.h"
+
+#include "FWCore/Framework/interface/EventSetup.h"
+//#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+
+#include "getBestVertex.h"
+
+#include "TrackingTools/Records/interface/TfGraphRecord.h"
+#include "PhysicsTools/TensorFlow/interface/TensorFlow.h"
+#include "RecoTracker/FinalTrackSelectors/interface/TfGraphDefWrapper.h"
+
+struct TfDnnCache {
+  TfDnnCache() : session_(nullptr) {}
+  tensorflow::Session* session_;
+};
+
+namespace {
+  class TfDnn {
+  public:
+    TfDnn(const edm::ParameterSet& cfg)
+        : tfDnnLabel_(cfg.getParameter<std::string>("tfDnnLabel"))
+
+    {
+      cache_ = new TfDnnCache();
+    }
+    TfDnnCache* cache_;
```

Comment:
```diff
- TfDnnCache* cache_;
+ tensorflow::Session* session_;
```
see the comment above about removing the struct `TfDnn`

Reply:
Hello. After I remove the TfDnnCache() struct and use a bare tensorflow session, there is the error "An exception of category 'InvalidSession' occurred while". I am not sure the reason but I guess this is relate to the memory management since in the old version, the pointer cache_ is managed by new and delete, and I cannot do the same thing for the bare tensorflow::session.

Reply:
Memory is still managed by `tensorflow::createSession` and `tensorflow::closeSession` in both cases, adding a struct that holds the pointer doesn't change the situation. Probably you have a coding error. Try the following, hope it helps:

```diff
diff --git a/RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc b/RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc
index 1aae133f52e..fc9cd9fa933 100644
--- a/RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc
+++ b/RecoTracker/FinalTrackSelectors/plugins/TrackTfClassifier.cc
@@ -22,16 +22,12 @@ namespace {
   public:
     TfDnn(const edm::ParameterSet& cfg)
         : tfDnnLabel_(cfg.getParameter<std::string>("tfDnnLabel"))
-
     {
-      cache_ = new TfDnnCache();
     }
-    TfDnnCache* cache_;
 
     ~TfDnn() {
-      if (cache_->session_) {
-        tensorflow::closeSession(cache_->session_);
-        delete cache_;
+      if (session_ != nullptr) {
+        tensorflow::closeSession(session_);
       }
     }
 
@@ -44,14 +40,12 @@ namespace {
     void beginStream() {}
 
     void initEvent(const edm::EventSetup& es) {
-      if (!cache_->session_) {
+      if (session_ == nullptr) {
         edm::ESHandle<TfGraphDefWrapper> tfDnnHandle;
         es.get<TfGraphRecord>().get(tfDnnLabel_, tfDnnHandle);
         tensorflow::GraphDef* graphDef_ = tfDnnHandle.product()->getGraphDef();
-        cache_->session_ = tensorflow::createSession(
-            graphDef_, 1);  //The integer controls how many threads are used for running inference
+        session_ = tensorflow::createSession(graphDef_, 1);  //The integer controls how many threads are used for running inference
       }
-      session_ = cache_->session_;
     }
 
     float operator()(reco::Track const& trk,
@@ -115,7 +109,7 @@ namespace {
     }
 
     const std::string tfDnnLabel_;
-    tensorflow::Session* session_;
+    tensorflow::Session* session_ = nullptr;
   };
 
   using TrackTfClassifier = TrackMVAClassifier<TfDnn>;
```